flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java

```diff
@@ -0,0 +1,345 @@
+			JobManagerMessages.TaskManagerInstance instance = (JobManagerMessages.TaskManagerInstance) Await.result(future, timeout);
+			Instance taskManager = instance.instance().get();
+
+			Future<Object> isBlobServiceDefined = taskManager.getActorGateway().ask(TaskManagerMessages.getIsBlobServiceDefined(), timeout);
+
+			if (!(Boolean) Await.result(isBlobServiceDefined, timeout)) {
+				display(ctx, request, "BlobService unavailable, cannot upload TaskManager logs.");
+				return;
+			}
+
+			//send log request to taskmanager
+			Future<Object> blobKeyFuture = taskManager.getActorGateway().ask(serveLogFile ? TaskManagerMessages.getRequestTaskManagerLog() : TaskManagerMessages.getRequestTaskManagerStdout(), timeout);
+			blobKeyFuture.onComplete(new OnComplete<Object>() {
+				@Override
+				public void onComplete(Throwable failure, Object success) throws Throwable {
+					try {
+						BlobKey blobKey = (BlobKey) success;
+
+						//delete previous log file, if it is different than the current one
+						if ((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).containsKey(taskManagerID)) {
+							if (!blobKey.equals((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).get(taskManagerID))) {
+								jobManager.tell(JobManagerMessages.getDeleteTaskManagerLog((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).remove(taskManagerID)));
+								(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+							}
+						} else {
+							(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+						}
+
+						//send blobkey to jobmanager
+						Future<Object> logPathFuture = jobManager.ask(JobManagerMessages.getRequestTaskManagerLog(blobKey), timeout);
+						String filePath = (String) Await.result(logPathFuture, timeout);
+
+						File file = new File(filePath);
+						final RandomAccessFile raf;
+						try {
+							raf = new RandomAccessFile(file, "r");
+						} catch (FileNotFoundException e) {
+							sendError(ctx, NOT_FOUND);
+							return;
+						}
+						long fileLength = raf.length();
+
+						HttpResponse response = new DefaultHttpResponse(HTTP_1_1, OK);
+						setContentTypeHeader(response, file);
+
+						if (HttpHeaders.isKeepAlive(request)) {
+							response.headers().set(CONNECTION, HttpHeaders.Values.KEEP_ALIVE);
+						}
+						HttpHeaders.setContentLength(response, fileLength);
+
+						// write the initial line and the header.
+						ctx.write(response);
+
+						// write the content.
+						ctx.write(new DefaultFileRegion(raf.getChannel(), 0, fileLength), ctx.newProgressivePromise())
+							.addListener(new GenericFutureListener<io.netty.util.concurrent.Future<? super Void>>() {
+								@Override
+								public void operationComplete(io.netty.util.concurrent.Future<? super Void> future) throws Exception {
+									synchronized (lastRequestCompleted) {
+										lastRequestCompleted.put(taskManagerID, true);
```

Comment:
What if a failure occurs before this listener is added. Wouldn't that mean that the `lastRequestCompleted` entry for this `taskManagerID` would remain `false` for all eternity?